### PR TITLE
Add alipay provider

### PR DESCRIPTION
**Pull request by leeqvip**
Provider for https://www.alipay.com/
Developer documentation: https://opendocs.alipay.com/open/284/web

Thank you! :)

### CODE REVIEW

File: src/Alipay/Provider.php [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+                RequestOptions::HEADERS     => [
+                    'content-type' => 'application/x-www-form-urlencoded;charset=utf-8',
+                ],
+            ]
+        );
+
+        $response = json_decode($response->getBody()->getContents(), true);
+
+        if (!empty($response['error_response']) || empty($response['alipay_user_info_share_response'])) {
+            throw new \InvalidArgumentException('You have error! '.\json_encode($response, JSON_UNESCAPED_UNICODE));
```
Comment (atymic) [inline src/Alipay/Provider.php:101]:
```suggestion
            throw new \InvalidArgumentException('Error getting apipay user details '.\json_encode($response, JSON_UNESCAPED_UNICODE));
```

Can we have a more helpful message here please?

File: src/Alipay/Provider.php [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+                RequestOptions::FORM_PARAMS => $this->getTokenFields($code),
+                RequestOptions::HEADERS     => [
+                    'content-type' => 'application/x-www-form-urlencoded;charset=utf-8',
+                ],
+            ]
+        );
+        $response = json_decode($response->getBody()->getContents(), true);
+
+        if (!empty($response['error_response'])) {
+            throw new \InvalidArgumentException('You have error! '.json_encode($response, JSON_UNESCAPED_UNICODE));
```
Comment (atymic) [inline src/Alipay/Provider.php:191]: Also here

File: src/Alipay/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+];
+```
+
+### Usage
+
+You should now be able to use the provider like you would regularly use Socialite (assuming you have the facade installed):
+
+```php
+return Socialite::driver('alipay')->redirect();
+```
```
Comment (atymic) [inline src/Alipay/README.md:43]: Can wee add the returned user fields please

File: src/Alipay/composer.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    "license": "MIT",
+    "authors": [
+        {
+            "name": "JonLee",
+            "email": "leeqvip@gmail.com"
+        }
+    ],
+    "require": {
+        "php": "^7.4 || ^8.0",
+        "ext-json": "*",
```
Comment (atymic) [inline src/Alipay/composer.json:20]: need to require openssl extension here.

**Review by atymic — CHANGES_REQUESTED** [review]

**Comment by atymic** [comment]
ping @leeqvip

File: src/Alipay/Provider.php [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+     * The scopes being requested.
+     *
+     * @var array
+     */
+    protected $scopes = ['auth_user'];
+
+    protected $baseUrl = 'https://openapi.alipay.com/gateway.do';
+    protected $authUrl = 'https://openauth.alipay.com/oauth2/publicAppAuthorize.htm';
+
+    protected $apiVersion = '1.0';
```
[the remark this answers is not in the document]
Comment (leeqvip) [reply src/Alipay/Provider.php:27]: They may be changed in the constructor. I don't want them to be defined as constants.

File: src/Alipay/composer.json [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    ],
+    "license": "MIT",
+    "authors": [
+        {
+            "name": "JonLee",
+            "email": "leeqvip@gmail.com"
+        }
+    ],
+    "require": {
+        "php": "^7.4 || ^8.0",
```
Comment (atymic) [inline src/Alipay/composer.json:19]:
```suggestion
        "php": "^8.0",
```

**Comment by atymic** [comment]
@lucasmichot can you give this a review as well please?

**Comment by atymic** [comment]
ping @leeqvip

File: src/Alipay/Provider.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (lucasmichot) [reply src/Alipay/Provider.php:27]: This is not the way we override environment related variables

File: src/Alipay/Provider.php [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+            $this->baseUrl,
+            [
+                RequestOptions::FORM_PARAMS => $params,
+            ]
+        );
+
+        $response = json_decode($response->getBody()->getContents(), true);
+
+        if (!empty($response['error_response']) || empty($response['alipay_user_info_share_response'])) {
+            throw new \InvalidArgumentException('Error getting apipay user details '.\json_encode($response, JSON_UNESCAPED_UNICODE));
```
Comment (lucasmichot) [inline src/Alipay/Provider.php:98]: Please import `InvalidArgumentException`

File: src/Alipay/Provider.php [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    }
+
+    /**
+     * Get the GET parameters for the code request.
+     *
+     * @param string|null $state
+     *
+     * @return array
+     */
+    protected function getCodeFields($state = null)
```
Comment (lucasmichot) [inline src/Alipay/Provider.php:127]: Why not just using most of the parent function?

File: src/Alipay/Provider.php [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+                RequestOptions::FORM_PARAMS => $this->getTokenFields($code),
+                RequestOptions::HEADERS     => [
+                    'content-type' => 'application/x-www-form-urlencoded;charset=utf-8',
+                ],
+            ]
+        );
+        $response = json_decode($response->getBody()->getContents(), true);
+
+        if (!empty($response['error_response'])) {
+            throw new \InvalidArgumentException('Error getting apipay access token '.json_encode($response, JSON_UNESCAPED_UNICODE));
```
Comment (lucasmichot) [inline src/Alipay/Provider.php:188]: Please import

File: src/Alipay/Provider.php [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+            'method'    => $method,
+            'timestamp' => date('Y-m-d H:m:s'),
+            'version'   => $this->apiVersion,
+        ];
+    }
+
+    /**
+     * @see https://opendocs.alipay.com/open/289/105656
+     */
+    protected function generateSign($params)
```
Comment (lucasmichot) [inline src/Alipay/Provider.php:217]: generateSign should contain the logic of `signWithSHA256RSA` and `buildParams`

File: src/Alipay/Provider.php [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+
+        openssl_sign($signContent, $signValue, $key, OPENSSL_ALGO_SHA256);
+
+        return base64_encode($signValue);
+    }
+
+    public static function buildParams(array $params, bool $urlencode = false, array $except = ['sign'])
+    {
+        $param_str = '';
+        foreach ($params as $k => $v) {
```
Comment (lucasmichot) [inline src/Alipay/Provider.php:242]: No need to reinvent `http_build_query` here

File: src/Alipay/README.md [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+Please see the [Base Installation Guide](https://socialiteproviders.com/usage/), then follow the provider specific instructions below.
+
+### Add configuration to `config/services.php`
+
+```php
+'alipay' => [
+  'client_id' => env('ALIPAY_APP_ID'),
+  'client_secret' => env('ALIPAY_RSA_PRIVATE_KEY'),
+  'redirect' => env('ALIPAY_REDIRECT'),
+  "sandbox" => env('ALIPAY_SANDBOX') == 'true'
```
Comment (lucasmichot) [inline src/Alipay/README.md:18]:
```suggestion
  'sandbox' => (bool) env('ALIPAY_SANDBOX', false),
```

File: src/Alipay/composer.json [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            "name": "JonLee",
+            "email": "leeqvip@gmail.com"
+        }
+    ],
+    "require": {
+        "php": "^8.0",
+        "ext-json": "*",
+        "ext-openssl": "*",
+        "socialiteproviders/manager": "~4.0"
```
Comment (lucasmichot) [inline src/Alipay/composer.json:22]:
```suggestion
        "socialiteproviders/manager": "^4.2"
```

**Review by lucasmichot — CHANGES_REQUESTED** [review]

**Comment by atymic** [comment]
cc @leeqvip

**Comment by atymic** [comment]
cc @leeqvip

**Comment by atymic** [comment]
Closing due to inactivity

**Pull request closed without merging**